extensions/ql-vscode/src/common/discovery.ts

```diff
@@ -36,49 +46,48 @@ export abstract class Discovery<T> extends DisposableObject {
     // other change notifications that might be coming along. However, this would create more
     // latency in the common case, in order to save a bit of latency in the uncommon case.
 
-    if (this.discoveryInProgress) {
+    if (this.currentDiscoveryPromise !== undefined) {
       // There's already a discovery operation in progress. Tell it to restart when it's done.
       this.retry = true;
     } else {
       // No discovery in progress, so start one now.
-      this.discoveryInProgress = true;
-      this.launchDiscovery();
+      this.currentDiscoveryPromise = this.launchDiscovery();
     }
+    return this.currentDiscoveryPromise;
   }
 
   /**
    * Starts the asynchronous discovery operation by invoking the `discover` function. When the
    * discovery operation completes, the `update` function will be invoked with the results of the
    * discovery.
    */
-  private launchDiscovery(): void {
-    const discoveryPromise = this.discover();
-    discoveryPromise
-      .then((results) => {
-        if (!this.retry) {
-          // Update any listeners with the results of the discovery.
-          this.discoveryInProgress = false;
-          this.update(results);
-        }
-      })
+  private async launchDiscovery(): Promise<void> {
+    let results: T | undefined;
+    try {
+      results = await this.discover();
+    } catch (err) {
+      void extLogger.log(
+        `${this.name} failed. Reason: ${getErrorMessage(err)}`,
+      );
+      results = undefined;
+    }
 
-      .catch((err: unknown) => {
-        void extLogger.log(
-          `${this.name} failed. Reason: ${getErrorMessage(err)}`,
-        );
-      })
+    if (this.retry) {
+      // Another refresh request came in while we were still running a previous discovery
+      // operation. Since the discovery results we just computed are now stale, we'll launch
+      // another discovery operation instead of updating.
+      // Note that by doing this inside of `finally`, we will relaunch discovery even if the
+      // initial discovery operation failed.
+      this.retry = false;
+      await this.launchDiscovery();
+    } else {
+      this.currentDiscoveryPromise = undefined;
 
-      .finally(() => {
-        if (this.retry) {
-          // Another refresh request came in while we were still running a previous discovery
-          // operation. Since the discovery results we just computed are now stale, we'll launch
-          // another discovery operation instead of updating.
-          // Note that by doing this inside of `finally`, we will relaunch discovery even if the
-          // initial discovery operation failed.
-          this.retry = false;
-          this.launchDiscovery();
-        }
-      });
+      // If the discovery was successful, then update any listeners with the results.
+      if (results !== undefined) {
+        this.update(results);
+      }
```

Comment:
(May not be valid if implementing the previous suggestion): By doing this after setting `this.currentDiscoveryPromise = undefined;`, it means that waiting on the `refresh` or `getCurrentRefreshPromise` methods doesn't mean that the `update` method has finished. Is this intentional?

Reply:
That's a good point. That wasn't intentional. Thankfully the previous suggestion has already changed this behaviour so that `update` must finish first.